<a href="https://colab.research.google.com/github/CodeTunisia/InformatiqueMathSup2122/blob/main/td1/TD%201%20-%20Les%20bases%20de%20la%20programmation%20avec%20Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exercice 1 : À la découverte des Variables (à faire dans la console)

In [ ]:
# Q1.
x = 5

Q2. Non on ne peut pas écrire l’égalité d’affectation dans le sens qu’on veut.

In [ ]:
5 = x

SyntaxError: can't assign to literal (<ipython-input-196-7f1a2e2f9b1f>, line 1)

In [ ]:
# Q3. 
y = 7 * x; print(y)

35


In [ ]:
x = 2; print(y)

35


In [ ]:
# Q4.
help(id)

Help on built-in function id in module builtins:

id(obj, /)
    Return the identity of an object.
    
    This is guaranteed to be unique among simultaneously existing objects.
    (CPython uses the object's memory address.)



In [ ]:
help(type)

Help on class type in module builtins:

class type(object)
 |  type(object_or_name, bases, dict)
 |  type(object) -> the object's type
 |  type(name, bases, dict) -> a new type
 |  
 |  Methods defined here:
 |  
 |  __call__(self, /, *args, **kwargs)
 |      Call self as a function.
 |  
 |  __delattr__(self, name, /)
 |      Implement delattr(self, name).
 |  
 |  __dir__(self, /)
 |      Specialized __dir__ implementation for types.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __init__(self, /, *args, **kwargs)
 |      Initialize self.  See help(type(self)) for accurate signature.
 |  
 |  __instancecheck__(self, instance, /)
 |      Check if an object is an instance.
 |  
 |  __repr__(self, /)
 |      Return repr(self).
 |  
 |  __setattr__(self, name, value, /)
 |      Implement setattr(self, name, value).
 |  
 |  __sizeof__(self, /)
 |      Return memory consumption of the type object.
 |  
 |  __subclasscheck__(self, subclass, /)
 |     

In [ ]:
type(x)

int

In [ ]:
id(x)

94097720517408

In [ ]:
#Q5.
x =x + 1; id(x)

94097720517472

In [ ]:
# Q6.
x = 11
x +=1; print(x)
x -=1; print(x)
x //=2; print(x)
x %=2; print(x)
x /=2; print(x)
x *=22; print(x)

12
11
5
1
0.5
11.0


In [ ]:
#Q7.
x, y = 7, 9
x, y = y, x
print(x, y)

9 7


# Exercice 2 : Premiers affichages (à faire dans la console)

In [ ]:
# Q1
Bonjour = 0

In [ ]:
# Q2
print(Bonjour)

0


In [ ]:
print("Bonjour") # ou print('Bonjour')

Bonjour


In [ ]:
# Q3.
type(Bonjour)

int

In [ ]:
type('Bonjour')

str

In [ ]:
#Q4
print('Aujourd\'hui et Il dit: "Bonjour"') # \' est un carctère (') hors syntaxe python.

Aujourd'hui et Il dit: "Bonjour"


In [ ]:
print("\"Aujourd’hui\"") # \" est un carctère (") hors syntaxe python.

"Aujourd’hui"


In [ ]:
# Q5
print("Hello World!")

Hello World!


In [ ]:
print("Hello\nWorld!") # \n  est un caractère de retour à la ligne

Hello
World!


In [ ]:
# Q6 
x, y, z = 1, 2, 3
print(x, y, z, sep = ";")

1;2;3


In [ ]:
print(x, y, z, sep = "et", end = '.')

1et2et3.

In [ ]:
print(x, y, z, sep = "\n")

1
2
3


In [ ]:
# Q7.
print("x={}, y={}, z={}".format(x, y, z))

x=1, y=2, z=3


# Exercice 4 : Quiz multiplications

In [ ]:
# Quiz multiplication
import random as rd
## Q1)
a = rd.randint(1, 12)
## Q2)
b = rd.randint(1, 12)
## Q3)
print("combien vaut le produit {} * {} ?".format(a, b))
## Q4)
p = int(input())
## Q5)
if p == a*b:
    print("Bravo!")
else:
    print("Perdu! La bonne réponse était ",a*b)

combien vaut le produit 8 * 5 ?
40
Bravo!


# Exercice 5 : Structures itératives
Le programme affiche un triangle d'étoiles.

In [ ]:
n = int(input("Entrez le nombre de ligne : "))
for i in range(n) :
    for j in range(i+1) :
        print(" ∗ ", end = "")
    print("\r")

Entrez le nombre de ligne : 5
 ∗ 
 ∗  ∗ 
 ∗  ∗  ∗ 
 ∗  ∗  ∗  ∗ 
 ∗  ∗  ∗  ∗  ∗ 


# Exercice 6 : Aire du rectangle par rapport au cercle

In [ ]:
from math import pi
r = 10.6
a = 1.3    # un côté du rectangle
aire_cercle = pi*r**2
b = 0      # valeur de départ choisie pour l'autre côté du rectangle
while a*b < aire_cercle:
    b += 1
b -= 1     # doit annuler la dernière mise à jour pour obtenir la bonne valeur
print("La plus grande valeur possible de b : ", b)

La plus grande valeur possible de b :  271


# Exercice 7 : Les fonctions

In [ ]:
import math
def ecart(a, b, epsilon = 0.1) :
    d = math.fabs(a - b)
    if (d < epsilon) :
        d = 0
    return d

In [ ]:
#Q1
print(ecart(a=12.2, b=11.9, epsilon = 1))

0


In [ ]:
print(ecart(a=12.2, b=11.9))

0.29999999999999893


In [ ]:
e= ecart(a=12.2, b=11.9) ;print(e)

0.29999999999999893


In [ ]:
# Q2.
import math
def ecart(a, b, epsilon = 0.1) :
    d = math.fabs(a - b)
    if (d < epsilon) :
        d = 0
    print(d)

In [ ]:
e1=ecart(a=12.2, b=11.9) ;print(e1)

0.29999999999999893
None


`0.29999999999999893` provient de l'affichage avec `print` dans `ecart()`.

`print(e1)` affiche la valeur `None` car la fonction `ecart()` ne retourne rien.


In [ ]:
ecart(a=12.2, b=11.9)

0.29999999999999893


# Exercice 8 : date plus un jour

## Q1. Déterminer si une année est bissextile ou non?

In [ ]:

def bissextile(a):
    return (a % 4 == 0 and a %100 != 0) or a % 400 ==0

In [ ]:
bissextile(2021)

False

In [ ]:
bissextile(2020)

True

## Q2. Donner le triplet (jour, mois et année) et affiche la date suivante.
* Les mois numéros 1, 3, 5, 7, 8, 10 et 12 ont 31 jours.
* Ceux de numéros 4, 6, 9 et 11 en ont 30.
* Le mois numéro 2 a 29 jours si l’année est bissextile, 28 dans le cas contraire.

### Une méthode

In [ ]:
def plus_un_jour(j, m, a):
    if m == 1 or m == 3 or m == 5 or m ==7 or m == 8 or m == 10 :
        if j == 31:
            j, m, a = 1, m+1,a
        else:
            j, m, a = j+1, m, a
    elif m == 12 :
        if j == 31:
            j, m, a = 1, 1, a+1
        else:
            j, m, a = j+1, m, a
    elif m == 4 or m == 6 or m == 9 or m == 11:
        if j == 30:
            j, m, a = 1, m+1,a
        else:
            j, m, a = j+1, m, a
    
    elif (j==28 and m==2 and not bissextile(a)) or (j==29 and m==2 and bissextile(a)):
        j, m, a = 1, m+1,a
    # La date à retouner par la fonction doit être valide!
    elif (m == 2 and not bissextile(a)and j>=29)or (m == 2 and bissextile(a)and j>29):
        return "Date invalide!"
    else :
        j, m, a = j+1, m, a
    # La date à retouner par la fonction doit être valide!
    if (1<= j <= 31 and 1 <= m <=12 and a >=1):
        return j, m, a
    else:
        return "Date invalide!"

In [ ]:
 plus_un_jour(31, 12, 2015)

(1, 1, 2016)

In [ ]:
 plus_un_jour(31, 6, 2019)

'Date invalide!'

In [ ]:
plus_un_jour(30, 4, 2019)

(1, 5, 2019)

In [ ]:
plus_un_jour(30, 3, 2020)

(31, 3, 2020)

In [ ]:
plus_un_jour(29, 2, 2021)

'Date invalide!'

In [ ]:
plus_un_jour(29, 2, 2020)

(1, 3, 2020)

### Une autre méthode
Une solution plus simple est possible en définissant une fonction intermédiaire qui renvoie le nombre de jours pour un mois et une année donnés.

In [ ]:
def nbrJours(m, a):
    if m == 2:
        if bissextile(a):
            nbr = 29
        else:
            nbr = 28
    elif m <= 7:
        if m % 2 == 1:
            nbr = 31
        else:
            nbr = 30
    else:
        if m % 2 == 0:
            nbr = 31
        else:
            nbr = 30
    return nbr

In [ ]:
nbrJours(2, 2021)

28

In [ ]:
def Plus_Un_Jour(j, m, a):
    if j < nbrJours(m, a): # si j avant le dernier jour
        j += 1
    else: # si j est le dernier jour
        j = 1 # le lendemain est le premier jour
        if m < 12: # du nouveau mois
            m += 1
        else: # de la nouvelle année
            m = 1
            a += 1
    return j, m, a

In [ ]:
Plus_Un_Jour(31, 12, 2015)

(1, 1, 2016)

In [ ]:
Plus_Un_Jour(28, 2, 2020)

(29, 2, 2020)

In [ ]:
Plus_Un_Jour(28, 2, 2021)

(1, 3, 2021)

# Exercice 9 : Nombres entiers

In [ ]:
# Q1
def inf(n):
    i = 0
    while i < n:
        print(i)
    i+=1
# Q2
def inf_pair(n):
    i = 0
    while i < n:
        if i%2==0:
            print(i)
    i+=1
# Q3          
def div(n):
    i = 1
    while i <=n:
        if n%i ==0 :
            print(i)
        i+=1
# Q4
def ndiv(n):
    ndiv = 0
    i = 1
    while i <=n:
        if n%i ==0 : # si i divise n, on ajoute 1 à ndiv
            ndiv+=1
        i+=1
    return ndiv
# Q5
def somme_div(n):
    s = 0
    i = 1
    while i <=n:
        if n%i ==0 : # si i divise n, on l'joute à s
            s+=i
        i+=1
    return s
# Q6 # Q7
def MaxMin(n):
    i = 0
    Max=1
    Min = Max
    posMin, posMax = 0, 0
    while i<=n:
        x = int(input("Donnez un entier :"))
        if x>Max:
            Max = x
            posMax = i
        elif x<=Min:
            Min = x
            posMin = i
        i+=1
    return (Max, posMax), (Min, posMin)
# Q8.
def MaxMin2():
    i = 0
    Max=1
    Min = Max
    posMin, posMax = 0, 0
    while True:
        x = int(input("Donnez un entier:"))
        if x>Max:
            Max = x
            posMax = i
        elif x<=Min:
            Min = x
            posMin = i
        if x == 0:
            break
        i+=1
    return (Max, posMax), (Min, posMin)

In [ ]:
MaxMin(5)

Donnez un entier :1
Donnez un entier :-2
Donnez un entier :8
Donnez un entier :9
Donnez un entier :7
Donnez un entier :3


((9, 3), (-2, 1))

In [ ]:
MaxMin2()

Donnez un entier:8
Donnez un entier:1
Donnez un entier:-2
Donnez un entier:0


((8, 0), (-2, 2))